game.sy

```diff
@@ -238,6 +270,95 @@ update :: fn delta: float -> void {
         p.try_grab' puck
     }
 
+    iter := 1000
+    t := 1.0
+    loop t > 0.0 && iter > 0 {
+        iter -= 1
+        is_collision := false
+        next_col := collision.Manifold {
+            t: 2.0,
+            normal: (0.0, 0.0),
+            a: collision.Wall {
+                pos: (0.0, 0.0),
+                normal: (0.0, 0.0),
+            },
+            b: collision.Wall {
+                pos: (0.0, 0.0),
+                normal: (0.0, 0.0),
+            },
+        }
+
+        // Find next collision
+        a_i := 0
+        // TODO(ed): Write this with for_each, when the upvalue bug is fixed.
+        loop a_i < len(bodies) {
+            a := bodies[a_i]
+            b_i := -1
+            bodies -> for_each' fn b: collision.CircleBody {
+                b_i += 1
+                // Skip the ones we've seen
+                if a_i >= b_i {
+                    ret
+                }
+                if a is :Puck || b is :Puck {
+                    ret
+                }
+                col :: collision.circle_vs_circle' a, b, delta * t * (a.vel - b.vel)
+                if col.t <= 0.0 || next_col.t < col.t {
+                    ret
+                }
+                next_col = col
+                is_collision = true
+            }
+            walls -> for_each' fn w: collision.Wall {
+                col :: collision.circle_vs_wall' a, w, delta * t * a.vel
+                if col.t <= 0.0 || next_col.t < col.t {
+                    ret
+                }
+                next_col = col
+                is_collision = true
+            }
+            corners -> for_each' fn c: collision.Circle {
+                col :: collision.circle_vs_circle' a, c, delta * t * a.vel
+                if col.t <= 0.0 || next_col.t < col.t {
+                    ret
+                }
+                if sign(col.normal[1]) != sign(c.pos[1]) {
+                    ret
+                }
+                next_col = col
+                is_collision = true
+            }
+            a_i += 1
+        }
+
+        // Step up to the next collision
+        step := min' t, next_col.t
+        bodies -> for_each' fn c: collision.Body {
+            c.pos += c.vel * step * delta
+        }
+
+        true <=> :collision.Body is :Player
+
+        // Solve the collision
+        if is_collision {
+            energy_kept :: 0.7
+            n :: next_col.normal
+            if :collision.Body is next_col.a {
+                a :!collision.CircleBody = next_col.a
+                a.pos += n * 0.01
+                a.vel = energy_kept * reflect' a.vel, n
+            }
+            if :collision.Body is next_col.b {
+                b :!collision.CircleBody = next_col.b
+                b.pos -= n * 0.01
+                b.vel = energy_kept * reflect' b.vel, n
+            }
+        }
+        t -= step
+    }
+    if iter == 0 { <!> }
```

Comment:
This shouldn't be in the final game - but I put it here so we see when we run out of collisions.